OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/TypeIterator.scala

```diff
@@ -1082,40 +1146,55 @@ class CFA_k_l_TypeIterator(project: SomeProject, val k: Int, val l: Int)
-                    case UBPS(tac: TheTACAI, isFinal) =>
-                        if (!isFinal)
-                            state.addDependency((depender, access._1, access._2), eOptP)
-
-                        val theTAC = tac.theTAC
-                        access._2.foldLeft(result) { (result, pc) =>
-                            theTAC.stmts(theTAC.properStmtIndexForPC(pc)).asPutField.objRef.asVar.definedBy.foldLeft(result) { (result, defSite) =>
-                                val defPC = if (defSite < 0) defSite else theTAC.stmts(defSite).pc
-                                combine(
-                                    result,
-                                    typesProperty(field, DefinitionSite(access._1, defPC), depender, NoContext, theTAC.stmts) // TODO Must actually supply valid context here!
-                                )
-                            }
-                        }
+        if (field.isDefinedField) {
+            if (field.definedField.isStatic) {
+                currentPointsTo(depender, field)
+            } else {
+                var result = emptyPointsToSet;
+                for {
+                    // Extract FieldWriteAccessInformation
+                    fai <- extractPropertyUB(
+                        EPK(field, FieldWriteAccessInformation.key), state.addDependency(depender, _)
+                    )
+                    (accessContextId, _, receiver, _) <- fai.accesses
+                    if receiver.isDefined
+
+                    // Extract TAC
+                    definedMethod = contextFromId(accessContextId).method.asDefinedMethod
+                    method = definedMethod.definedMethod
+                    tacEP <- extractPropertyUB(
+                        EPK(method, TACAI.key), state.addDependency((depender, definedMethod, receiver), _)
+                    )
+                    theTAC <- tacEP.tac
+                    defSite <- uVarForDefSites(receiver.get, theTAC.pcToIndex).definedBy
 
-                    case _ =>
-                        state.addDependency((depender, access._1, access._2), eOptP)
-                        result
+                    // Extract caller context
+                    callers <- extractPropertyUB(
+                        EPK(definedMethod, Callers.key),
+                        state.addDependency((depender, definedMethod, receiver), _)
+                    )
+                    (calleeContext, _, _, _) <- callers.callContexts(definedMethod)(this).iterator
+                } {
+                    val defPC = if (defSite < 0) defSite else theTAC.stmts(defSite).pc
+
+                    result = combine(
+                        result,
+                        typesProperty(field, DefinitionSite(method, defPC), depender, calleeContext, theTAC.stmts)
+                    )
                 }
+
+                result
             }
+        } else {
+            NoAllocationSites // FIXME: Deal with VirtualDeclaredFields
```

Comment:
I guess (guessing as i did not place this FIXME myself) we could return `currentPointsTo` in case the field is defined and static. Information about fields being static is not yet persisted for VirtualDeclaredFields, else we could generalize this.
The else part (currently: the field is defined but an instance field) we could probably just also allow for virtual declared fields.